File: src/2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
#time defeat: 76%
#space defeat: 5%
#总体想法-分治，逐步向下
class Solution:
    def addTwoNodes(self, _resultNode, _node1, _node2, _carry):
	#终止情况
        if _node1 == None and _node2 == None and _carry == 0:
            return None
	#当两链表不对齐时，补齐链表
        if not _node1:
            _node1 = ListNode(0, None)
        if not _node2:
            _node2 = ListNode(0, None)
	#构造新节点
        val = _node1.val + _node2.val + _carry
        newCarry = 0
        if val >= 10:
            newCarry = val // 10
            _resultNode.val = val - newCarry * 10
        else:
            _resultNode.val = val
        #print(_node1.val, _node2.val, _resultNode.val)
        newNode = ListNode(0, None)
        _resultNode.next = self.addTwoNodes(newNode, _node1.next, _node2.next, newCarry)
        return _resultNode
        
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
	#极端情况
        if l1.val == 0 and l1.next == None:
            return l2
        elif l2.val == 0 and l2.next == None:
            return l1
        else:
	    #普遍情况
            newNode = ListNode(0, None)
            return self.addTwoNodes(newNode, l1, l2, 0)
        

                
                
        

File: src/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        //想要做一种偷鸡做法，但是一看长度最长为100，偷鸡还得转成字符串
        //就有点麻烦了
        //干翻腾讯CSIG
        //还是模拟把
        //注意进位，注意不等长情况
        //边界处理，虽然看起来没用
        //二十分钟，两次提交，时间击败-45.15%，空间击败-18.44%
        if(l1 == nullptr){
            return l2;
        }
        if(l2 == nullptr){
            return l1;
        }
        int carry = 0;  //进位
        ListNode * p1 = l1;
        ListNode * p2 = l2; //遍历指针
        //需要新建链表
        ListNode * fakeHead = new ListNode(0);
        ListNode * temp = fakeHead;
        //当两个都有的时候
        while(p1 && p2){
            int sum = p1->val + p2->val + carry;
            carry = sum / 10;   //更新进位
            sum %= 10;  //去掉进位
            //新建节点
            ListNode * node = new ListNode(sum);
            //连入链表
            temp->next = node;
            temp = node;
            //推进指针
            p1 = p1->next;
            p2 = p2->next;
        }
        //当一个链表推进完时
        //让p1指向未完链表
        p1 = (p2 == nullptr) ? p1 : p2;
        //看看有没有进位
        if(carry == 0){
            //没有的话，直接连入
            temp->next = p1;
        }
        else{
            //有进位，处理
            while(carry && p1){
                int sum = p1->val + carry;
                carry = sum / 10;
                sum %= 10;
                ListNode * node = new ListNode(sum);
                temp->next = node;
                temp = node;
                p1 = p1->next;
            }
            //如果还有进位
            if(carry){
                ListNode * node = new ListNode(carry);
                temp->next = node;
                temp = node;    
            }
            else{
                //如果没有了
                temp->next = p1;
            }
        }
        return fakeHead->next;
    }
};
